mypy/plugins/dataclasses.py

```diff
@@ -769,3 +770,63 @@ def _is_dataclasses_decorator(node: Node) -> bool:
     if isinstance(node, RefExpr):
         return node.fullname in dataclass_makers
     return False
+
+
+def replace_function_sig_callback(ctx: FunctionSigContext) -> CallableType:
+    """
+    Generates a signature for the 'dataclasses.replace' function that's specific to the call site
+    and dependent on the type of the first argument.
+    """
+    if len(ctx.args) != 2:
+        # Ideally the name and context should be callee's, but we don't have it in FunctionSigContext.
+        ctx.api.fail(f'"{ctx.default_signature.name}" has unexpected type annotation', ctx.context)
+        return ctx.default_signature
+
+    if len(ctx.args[0]) != 1:
+        return ctx.default_signature  # leave it to the type checker to complain
+
+    inst_arg = ctx.args[0][0]
+
+    # <hack>
+    from mypy.checker import TypeChecker
+
+    assert isinstance(ctx.api, TypeChecker)
+    obj_type = ctx.api.expr_checker.accept(inst_arg)
+    # </hack>
+
+    obj_type = get_proper_type(obj_type)
+    if not isinstance(obj_type, Instance):
+        return ctx.default_signature
+    inst_type_str = format_type_bare(obj_type)
+
+    metadata = obj_type.type.metadata
+    dataclass = metadata.get("dataclass")
+    if not dataclass:
+        ctx.api.fail(
+            f'Argument 1 to "replace" has incompatible type "{inst_type_str}"; expected a dataclass',
+            ctx.context,
+        )
+        return ctx.default_signature
+
+    arg_names = [None]
+    arg_kinds = [ARG_POS]
+    arg_types = [obj_type]
+    for attr in dataclass["attributes"]:
```

Comment:
I wish I could've deserialized metadata, or even accessed a list of `DataclassAttribute`s that's already been deserialized...

Or maybe the right way to go is to replace `dataclasses.replace` with an `OverloadedFuncDef` that the `transform` (which has full context) will keep adding to.

Reply:
The `OverloadedFuncDef` appears to be a non-starter since by the time I can do anything to replace `dataclasses.replace` with a synthesized `OverloadedFuncDef`, the call expression has been resolved with the pointer to the generic `FunctionDef`.

Reply:
OK, solved it by creating a "secret" symbol with a signature at semantic analysis time, then using this signature in the function sig callback.